Add support for color animated properties

nc4rrillo opened the pull request:
Fixes: #2118 

Hi,
This is a patch to add support for animated color properties in WebRender.

### Some things to note:
- I modified the animations sample to show off the color animations. I hope my change is okay :).
- There is an assertion that popped up as a result of me calling `gpu_cache.invalidate()` in `prepare_prim_for_render`. 

#### STR for the assertion is:
* Open animations sample
* Hit ',' or ',.' to rotate the rects until they go off screen.

### Things to do before merging:
- [ ] Fix assertion in `gpu_cache.rs` line 612
- [x] Decide on default value for color bindings
- [x] Investigate border glitch
- [x] Rebase
- [ ] Measure serialization/deserialization impact
- [ ] Bump WR API version

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/2149)
<!-- Reviewable:end -->

nc4rrillo commented:
Other stuff I've noticed:
Some parts of the code like `box_shadow.rs` when building the frame has a check in `add_box_shadow`:

```
        if color.a == 0.0 {
            return;
        }
```

which causes an issue with a animated color binding that's initially transparent. Since it never gets added to the frame, it never gets a primitive, and therefore even when the color binding is updated to something not completely transparent, the changes aren't picked up.

There's similar logic for rects for example.

nc4rrillo commented:
backtrace of the assertion mentioned in the PR description:
```
thread 'WRRenderBackend#0' panicked at 'assertion failed: `(left == right)`
  left: `Epoch(51)`,
 right: `Epoch(50)`', src/gpu_cache.rs:612:8
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:69
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:58
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:538
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: webrender::gpu_cache::GpuCache::get_address
             at src/gpu_cache.rs:612
   8: webrender::tiling::add_to_batch
             at src/tiling.rs:395
   9: webrender::tiling::<impl webrender::prim_store::PrimitiveRun>::add_to_batch
             at src/tiling.rs:72
  10: webrender::tiling::<impl webrender::picture::PicturePrimitive>::add_to_batch
             at src/tiling.rs:957
  11: webrender::tiling::add_to_batch
             at src/tiling.rs:804
  12: webrender::tiling::<impl webrender::prim_store::PrimitiveRun>::add_to_batch
             at src/tiling.rs:72
  13: webrender::tiling::<impl webrender::picture::PicturePrimitive>::add_to_batch
             at src/tiling.rs:957
  14: webrender::tiling::add_to_batch
             at src/tiling.rs:804
  15: webrender::tiling::<impl webrender::prim_store::PrimitiveRun>::add_to_batch
             at src/tiling.rs:72
  16: webrender::tiling::<impl webrender::picture::PicturePrimitive>::add_to_batch
             at src/tiling.rs:957
  17: webrender::tiling::add_to_batch
             at src/tiling.rs:804
  18: webrender::tiling::<impl webrender::prim_store::PrimitiveRun>::add_to_batch
             at src/tiling.rs:72
  19: webrender::tiling::<impl webrender::picture::PicturePrimitive>::add_to_batch
             at src/tiling.rs:957
  20: webrender::tiling::add_to_batch
             at src/tiling.rs:804
  21: webrender::tiling::<impl webrender::prim_store::PrimitiveRun>::add_to_batch
             at src/tiling.rs:72
  22: webrender::tiling::<impl webrender::picture::PicturePrimitive>::add_to_batch
             at src/tiling.rs:957
  23: webrender::tiling::add_to_batch
             at src/tiling.rs:804
  24: webrender::tiling::<impl webrender::prim_store::PrimitiveRun>::add_to_batch
             at src/tiling.rs:72
  25: webrender::tiling::<impl webrender::picture::PicturePrimitive>::add_to_batch
             at src/tiling.rs:957
  26: webrender::tiling::AlphaBatcher::build
             at src/tiling.rs:1036
  27: <webrender::tiling::ColorRenderTarget as webrender::tiling::RenderTarget>::build
             at src/tiling.rs:1407
  28: webrender::tiling::RenderPass::build
             at src/tiling.rs:1805
  29: webrender::frame_builder::FrameBuilder::build
             at src/frame_builder.rs:1790
  30: webrender::frame::FrameContext::build_rendered_document
             at src/frame.rs:1185
  31: webrender::render_backend::Document::render
             at src/render_backend.rs:117
  32: webrender::render_backend::RenderBackend::process_document
             at src/render_backend.rs:419
  33: webrender::render_backend::RenderBackend::run
             at src/render_backend.rs:491
  34: webrender::renderer::Renderer::new::{{closure}}
             at src/renderer.rs:1988
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: Other, error: StringError("cannot send on closed channel") }) }', /checkout/src/libcore/result.rs:906:4
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:69
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:58
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:538
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:498
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:71
   9: core::result::unwrap_failed
             at /checkout/src/libcore/macros.rs:23
  10: <core::result::Result<T, E>>::unwrap
             at /checkout/src/libcore/result.rs:772
  11: webrender_api::api::RenderApi::send
             at /home/nc4rrillo/Documents/projects/webrender/webrender_api/src/api.rs:565
  12: webrender_api::api::RenderApi::generate_frame
             at /home/nc4rrillo/Documents/projects/webrender/webrender_api/src/api.rs:760
  13: <animation::App as animation::boilerplate::Example>::on_event
             at examples/animation.rs:341
  14: animation::boilerplate::main_wrapper
             at examples/common/boilerplate.rs:264
  15: animation::main
             at examples/animation.rs:429
  16: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:99
  17: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:459
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  18: main
  19: __libc_start_main
  20: _start
```

glennw commented:
Wow, this is amazing work. Apologies I haven't gotten to it yet - will run through this tomorrow.

Not sure if relevant, but I think it's completely fine to do this incrementally (e.g. merge support for just animated color on rects, then follow up with other primitives etc), if that makes things easier.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #2146) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
Reviewed 23 of 23 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[webrender/examples/animation.rs, line 296 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-_-TpS-l2YnawXATD7:-L-_-TpS-l2YnawXATD8:b-3ru1) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/webrender/examples/animation.rs#L296)):*
> ```Rust
>                 y += 200.;
>             }
>             
> ```

nit: extra spaces

---

*[webrender/src/frame_builder.rs, line 844 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-_09HiponHb9RbK33o:-L-_09HiponHb9RbK33p:b-aaud2o) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/webrender/src/frame_builder.rs#L844)):*
> ```Rust
>             &PropertyBinding::Value(value) => PropertyBinding::Value(value.premultiplied()),
>             &PropertyBinding::Binding(binding) => {
>                 let property_binding_key : PropertyBindingKey<PremultipliedColorF> = unsafe { mem::transmute(binding.key) };
> ```

If we were storing the value outside the property, then that would probably make this a lot simpler and remove the transmute nastiness you mentioned?

---

*[webrender/src/prim_store.rs, line 1434 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-_013XXnOuXc7IqNRZ:-L-_013XXnOuXc7IqNR_:b-3ru1) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/webrender/src/prim_store.rs#L1434)):*
> ```Rust
>                 PrimitiveKind::Rectangle => {
>                     let mut rect = &mut self.cpu_rectangles[metadata.cpu_prim_index.0];
>                     
> ```

nit: extra spaces

---

*[wrench/src/rawtest.rs, line 201 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-ZzMi2_jys9Yx3WkpS:-L-ZzMi37PQ5vqd0qa_f:b-28wvjo) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/wrench/src/rawtest.rs#L201)):*
> ```Rust
>             builder.push_clip_id(clip);
>             builder.push_rect(&PrimitiveInfo::new(rect(100., 100., 100., 100.)),
>                               ColorF::new(0.0, 0.0, 1.0, 1.0).into());
> ```

Perhaps we could instead have the interface be something like:

```
fn push_rect<P>(&mut self, color: P) where P: Into<PropertyBinding<ColorF>>
```

So that you can push a color directly in the common case?

---

*[wrench/src/yaml_frame_writer.rs, line 739 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-_-4zR_We2A7nirT4y:-L-_-4zR_We2A7nirT4z:bdl63t5) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/wrench/src/yaml_frame_writer.rs#L739)):*
> ```Rust
>                     println!("TODO YAML YuvImage");
>                 }
>                 Border(mut item) => {
> ```

Having to use all the mutability on stored items is why I originally had the resolved value stored outside the property binding. There's pros and cons to each.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2149)*
<!-- Sent from Reviewable.io -->

glennw commented:
This generally looks good to me! My main questions are whether (a) storing the resolved value outside the binding would remove a lot of the extra mutability here and also fix the transmute in the PR, and (b) if we should use the ```Into``` trick mentioned above to simplify the case for callers that are just providing colors without a binding.

nc4rrillo commented:
@glennw that `Into` trait is a good idea. I agree with moving the resolved value out of  `PropertyBinding`. I'll make that change as well. 

I'm also going to revert my changes to add this animated color support to borders and save that for a follow up.

jrmuizel commented:
I'm a little worried about the additional serialization/deserialization and display list size cost this might add. Should we measure it?

glennw commented:
@jrmuizel Yea, I'll do some measurements before merging anything.

nc4rrillo commented:
I've rebased this again and implemented the changes you suggested.

The `Into<PropertyBinding<ColorF>>` is very good in that very little places need an API change. 
AFAICT it should only be places that use `GradientStops` since `create_gradient` takes a `Vec<GradientStop>` directly.

I also reverted my change to `PropertyBinding<T>` to remove some of the mutability in my earlier changes. 

For now only rects, box shadows, and gradients have color animation support. 

Besides the assertion, another open question I have is about what good behavior for an initial value for color bindings would be. Should the client be expected to `GenerateFrame` with the initial value? Or should I try to do something similar to what `FilterOp` does with opacity?

cc: @glennw

nc4rrillo commented on webrender/src/prim_store.rs:
```diff
@@ -1647,6 +1656,19 @@ impl PrimitiveStore {
+                    (None, true, true)
+                },
+                PrimitiveKind::AlignedGradient | PrimitiveKind::AngleGradient | PrimitiveKind::RadialGradient => {
+                    gpu_cache.invalidate(&metadata.gpu_location);
```
I don't like that I unconditionally invalidate here.

nc4rrillo commented on webrender/src/prim_store.rs:
```diff
@@ -1647,6 +1656,19 @@ impl PrimitiveStore {
+                PrimitiveKind::Brush => {
+                    let brush = &mut self.cpu_brushes[metadata.cpu_prim_index.0];
+                    if let BrushKind::Solid { ref mut color, ref binding } = brush.kind {
+                        *color = scene_properties.resolve_color(&binding);
```
Maybe I should add an `if let &PropertyBinding::Binding(_) = binding` here so I'm not trying to resolve `PropertyBinding::Value` properties?

nc4rrillo commented:
Review status: 0 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[webrender/examples/animation.rs, line 296 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-_-TpS-l2YnawXATD7:-L-_cMppGFrbu7fpiMC7:b-896fix) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/webrender/examples/animation.rs#L296)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

nit: extra spaces
</blockquote></details>

Done.

---

*[webrender/src/frame_builder.rs, line 844 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-_09HiponHb9RbK33o:-L-_bLHE14BB3dQ5HzBT:b-896fix) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/webrender/src/frame_builder.rs#L844)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

If we were storing the value outside the property, then that would probably make this a lot simpler and remove the transmute nastiness you mentioned?
</blockquote></details>

Done.

---

*[webrender/src/prim_store.rs, line 1434 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-_013XXnOuXc7IqNRZ:-L-_cO10M05hWLyHh1gQ:b-896fix) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/webrender/src/prim_store.rs#L1434)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

nit: extra spaces
</blockquote></details>

Done.

---

*[wrench/src/yaml_frame_writer.rs, line 739 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-_-4zR_We2A7nirT4y:-L-_bOov8QtQYYslaAG4:bdbtd37) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/wrench/src/yaml_frame_writer.rs#L739)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Having to use all the mutability on stored items is why I originally had the resolved value stored outside the property binding. There's pros and cons to each.
</blockquote></details>

Done. Reverted my change to PropertyBinding.

---

*[wrench/src/rawtest.rs, line 201 at r1](https://reviewable.io:443/reviews/servo/webrender/2149#-L-ZzMi2_jys9Yx3WkpS:-L-_bTZVBWEmgLHNKxr7:b-896fix) ([raw file](https://github.com/servo/webrender/blob/2539749c71579cd0b009a7d28d6995ede80018ce/wrench/src/rawtest.rs#L201)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Perhaps we could instead have the interface be something like:

```
fn push_rect<P>(&mut self, color: P) where P: Into<PropertyBinding<ColorF>>
```

So that you can push a color directly in the common case?

</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2149)*
<!-- Sent from Reviewable.io -->

nc4rrillo commented:
Accidentally closed. Reopening.

nc4rrillo commented:
@glennw so I managed to also add support for borders (including complex ones) by moving the GPU block building out of the border code and into the primitive code.

I did however run into the following issue while putting together the sample app where some of the borders are glitched:
![](https://i.imgur.com/O139u1u.png)

It only happens with Dashed/Dotted borders of a non zero radius.
I'm investigating if it has anything to do with my changes or if it existed previously

nc4rrillo commented:
Looks like the above is a regression unrelated to my change:
```
8922f5dc5c3c5f34de7a4ce2a0f3ca8de1821717 is the first bad commit
commit 8922f5dc5c3c5f34de7a4ce2a0f3ca8de1821717
Author: Ethan Lin <ethlin@mozilla.com>
Date:   Wed Nov 22 16:35:11 2017 +0800

    Add cache key for box shadow.
```

nc4rrillo commented:
So the glitch above isn't related to my change, I'm trying to track down exactly _what_ caused it but I'm pretty far back in the commit history and not having any luck so far. It seems to be affected by the allocation patterns of the render targets, as different changes (for example box shadow caching) have impacted where I see it show up.

nc4rrillo commented:
it gets weirder 🥇 
![](https://i.imgur.com/8TBcr2i.png)

glennw commented:
@nc4rrillo Interesting glitches! But if unrelated to this PR, it doesn't need to block this. Is this ready for re-review again, or is it still in progress?

nc4rrillo commented:
@glennw This is ready for review now. For default values, I hang it off of `PropertyBinding::Binding`, the only place where its mutated is in `picture.rs` where it works much like how `FilterOp` did.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #2091) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #2192) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
Could you remove the merge commits and rebase it onto master instead? I'll run through this again today or tomorrow, and try to do some measurements of the DL deserialization time differences. I don't expect anything major, but I'll check just in case.

glennw commented:
Reviewed 4 of 23 files at r2, 5 of 15 files at r3, 5 of 8 files at r4, 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[webrender/src/border.rs, line 111 at r5](https://reviewable.io:443/reviews/servo/webrender/2149#-L-srUoUrDI5KCP7E9Vh:-L-srUoUrDI5KCP7E9Vi:bh2s9b7) ([raw file](https://github.com/servo/webrender/blob/887e7c1253605eb2c3b610bb1f7ecac34bbbbd14/webrender/src/border.rs#L111)):*
> ```Rust
>     fn get_corner(
>         &self,
>         scene_properties: &SceneProperties,
> ```

I think we might need to leave doing borders to a follow up PR. In this code here, we determine how we are going to render the border during the scene construction. If the colors then change by an animation property, we would need to re-run this code to determine the most efficient path to render the border.

---

*[webrender/src/box_shadow.rs, line 64 at r5](https://reviewable.io:443/reviews/servo/webrender/2149#-L-sroM58x_FQ1T3XmOP:-L-sroM58x_FQ1T3XmOQ:bsq2bdz) ([raw file](https://github.com/servo/webrender/blob/887e7c1253605eb2c3b610bb1f7ecac34bbbbd14/webrender/src/box_shadow.rs#L64)):*
> ```Rust
>         clip_mode: BoxShadowClipMode,
>     ) {
>         let color = scene_properties.resolve_color(&binding);
> ```

Same issue here, that we resolve during scene construction. If we remove the box shadow, and then it gets animated to a valid color, it won't appear.

---

*[webrender/src/frame_builder.rs, line 764 at r5](https://reviewable.io:443/reviews/servo/webrender/2149#-L-ss0ZUTs6uwQxl7Wu5:-L-ss0ZUTs6uwQxl7Wu6:b1mifx7) ([raw file](https://github.com/servo/webrender/blob/887e7c1253605eb2c3b610bb1f7ecac34bbbbd14/webrender/src/frame_builder.rs#L764)):*
> ```Rust
>         aa_mode: BrushAntiAliasMode,
>     ) {
>         let color = scene_properties.resolve_color(&binding);
> ```

Same issue here, unfortunately. We can't discard rectangles based on the scene properties, since they will then be removed if the initial resolved color is invalid.

---

*[webrender/src/picture.rs, line 142 at r5](https://reviewable.io:443/reviews/servo/webrender/2149#-L-ssTsnCWD4PShE7g42:-L-ssTsoZmzQ4JPV5nOH:bwcnb3p) ([raw file](https://github.com/servo/webrender/blob/887e7c1253605eb2c3b610bb1f7ecac34bbbbd14/webrender/src/picture.rs#L142)):*
> ```Rust
>                             &mut FilterOp::Opacity(ref mut binding) => {
>                                 let value = properties.resolve_float(binding);
>                                 match binding {
> ```

I don't quite follow why we need this code here?

---

*[webrender/src/prim_store.rs, line 509 at r5](https://reviewable.io:443/reviews/servo/webrender/2149#-L-sscPs5sWfyfZJvhQ6:-L-sscPs5sWfyfZJvhQ7:b-3ru1) ([raw file](https://github.com/servo/webrender/blob/887e7c1253605eb2c3b610bb1f7ecac34bbbbd14/webrender/src/prim_store.rs#L509)):*
> ```Rust
>         request.push([self.widths.left,  self.widths.top, 
>                       self.widths.right, self.widths.bottom]);
>         
> ```

nit: extra spaces

---

*[webrender/src/scene.rs, line 33 at r5](https://reviewable.io:443/reviews/servo/webrender/2149#-L-sr1IKaif2EWvNL1qS:-L-sr1IKaif2EWvNL1qT:b-3ru1) ([raw file](https://github.com/servo/webrender/blob/887e7c1253605eb2c3b610bb1f7ecac34bbbbd14/webrender/src/scene.rs#L33)):*
> ```Rust
>         self.float_properties.clear();
>         self.color_properties.clear();
>         
> ```

nit: extra spaces

---

*[webrender/src/scene.rs, line 217 at r5](https://reviewable.io:443/reviews/servo/webrender/2149#-L-srDTl5yMXRMsIGS0i:-L-srDTl5yMXRMsIGS0j:b-3ru1) ([raw file](https://github.com/servo/webrender/blob/887e7c1253605eb2c3b610bb1f7ecac34bbbbd14/webrender/src/scene.rs#L217)):*
> ```Rust
>                     PropertyBinding::Binding(_, value) => value
>                 };
>                 
> ```

nit: extra spaces

---

*[webrender/src/tiling.rs, line 808 at r5](https://reviewable.io:443/reviews/servo/webrender/2149#-L-ssnwd86iCLKvgD29c:-L-ssnwd86iCLKvgD29d:b-2g33l1) ([raw file](https://github.com/servo/webrender/blob/887e7c1253605eb2c3b610bb1f7ecac34bbbbd14/webrender/src/tiling.rs#L808)):*
> ```Rust
>                                                 FilterOp::Brightness(amount) => (7, amount),
>                                                 FilterOp::Opacity(binding) => {
>                                                     let amount = match binding {
> ```

Perhaps we should have a method to return the current value, to avoid duplication.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/2149)*
<!-- Sent from Reviewable.io -->

glennw commented:
Ugh, I'm *so sorry* this has turned out to be way more complex than I thought it would. It hadn't occurred to me that we will need to change a lot of how we early reject primitives based on invalid color etc (which are what most of the substantive comments in the last review are in relation to).

I'll be traveling most of next week - but I'll try to follow up on this again. I wonder if it's worth trying to land this with initial support for just the API changes and rectangles only (with the fixes to early discard of the primitives mentioned above)? That might let us get the initial support landed, and then work incrementally on each other primitive. Thoughts?

glennw commented:
PS: I removed the ```easy``` tag from the bug this references :)

nc4rrillo commented:
@glennw That sounds good to me. I'll rework this to support only rectangles for now.

glennw commented:
@nc4rrillo Thanks, and again, apologies for how tricky this has turned out to be.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #2207) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
@nc4rrillo Should we close this for now?

nc4rrillo commented:
Yes I will reopen a PR once I have more time to work on this 🙂
